### PR TITLE
Fix mapped int negative or overflow bugs, improve color C API

### LINKED ISSUE #1408: Windows: PyLong_AsLong signedness causes integer conversion error when unmapped color int is passed to draw functions (2732)
#  **Issue &numero;2732 opened by *[bast0006](https://github.com/bast0006)* at 2021-09-25 07:37:59**

In CHECK_LOAD_COLOR pygame calls PyLong_Check (which is aliased from PyInt_Check), and then PyLong_AsLong (aliased as PyInt_AsLong). However, it's then immediately cast to a Uint32. PyLong_AsLong doesn't return an unsigned integer, it returns a signed integer, and all 32 bits are required to support certain RRGGBBAA octuplets such as white. 

https://github.com/pygame/pygame/blob/6eb04ef9fcb6dd37aaaa303c2c16daac70ddec8b/src_c/draw.c# L83

This should instead be calling PyLong_AsUnsignedLong to properly support all 32 bit colors. 

Minimal reproduction:

```py
import pygame
s = pygame.Surface((10, 10), pygame.SRCALPHA)
pygame.draw.line(s, 0xFFEEDDCC, (0, 0), (1, 1)) 
```
Expected result (and result on my OSX test machine:
`<rect<0, 0, 0, 0>>`

Actual result:
```
OverflowError: Python int too large to convert to C long

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
SystemError: <built-in function line> returned a result with an error set
```

pygame 2.0.1 (SDL 2.0.14, python 3.7.9). Was also tested and failed on 3.8.10, 3.9.5. 

Note that pygame.Color(0xFFFFFFFF) works as it goes through a different pathway, and is an effective workaround for this issue. Colors below 0x80000000 function normally on both platforms.

Thanks to DigiDuncan# 4979 for catching the error so I could debug and write it up.
<hr>

#  Comments

# # #  **[DigiDuncan](https://github.com/DigiDuncan)* commented at 2021-09-25 07:41:35*

Thanks so much for helping me diagnose this bug and comprehensively writing up the issue!
<hr>

# # #  **[Starbuck5](https://github.com/Starbuck5)* commented at 2021-09-25 18:31:40*

This is an impressive bug report!

And there's definitely a problem here, but I don't think it's exactly what you think.

The draw functions have the option of taking 1 integer, but the docs say:
> an integer value that has been mapped to the surface's pixel format (see pygame.Surface.map_rgb()convert a color into a mapped color value and pygame.Surface.unmap_rgb()convert a mapped integer color value into a Color)

So the reason that lower hex numbers work at all is kind of a coincidence, because the pygame.SRCALPHA puts the surfaces into the correct format for receiving these as mapped colors? Even that seems wrong, because when you run:

```py
surf = pygame.Surface((50,50), pygame.SRCALPHA)
print(0xFFEEDDCC)
print(surf.map_rgb(pygame.Color(0xFFEEDDCC)))
print(0x80000000)
print(surf.map_rgb(pygame.Color(0x80000000)))
```

You get:

```
4293844428
-855642403
2147483648
8388608
```

The mapped rgb (what the draw command thinks it's taking?) doesn't match the hex.

This may be sort of an edge case then, because in pygame 2 and above `draw` is supposed to take any valid `pygame.Color()` arguments, but the single int case already had a use?

This is more of undefined behavior than a bug per se, although there definitely is stuff to fix. Like this should have a better error message, and better error handling, and maybe a docs change is in order. For @DigiDuncan, I would suggest passing in the hex as a string, like `pygame.draw.line(surf, "# FFEEDDCC", (10,0), (50,50))` or `pygame.draw.line(surf, "0xFFEEDDCC", (10,0), (50,50))`.

I haven't been able to dig into this as far as I would like, but in conclusion, this is a complex issue :smile:
<hr>

# # #  **[bast0006](https://github.com/bast0006)* commented at 2021-09-26 01:31:36*

Ah, I see. I hadn't quite realized that was how it worked! I am more used to the 1.x way of doing things, and I've never directly passed an integer in here. I missed the note at the top of the draw docs that says how it behaves, but in hindsight that makes complete sense, thanks!

### LINKED ISSUE #2336: Passing an color int with high bit set into a drawing function crashes with an overflow error
**Environment:**

You can get some of this info from the text that pops up in the console when you run a pygame program.

- **Operating system** (e.g. Windows, Linux(Debian), Linux(Ubuntu), Mac): Windows 11
- **Python version** (e.g. 3.7.9, 3.8.5) : 3.11.4
- **SDL version** (e.g. SDL 2.0.12): 2.26.5
- **PyGame version** (e.g. 2.0.0.dev10, 1.9.6): 2.3.1.dev2
- **Relevant hardware** (e.g. if reporting a bug about a controller, tell us the brand & name of it): N.A.

**Current behavior:**

setting the high bit in an `int` put into a drawing function bombs with an overflow error

**Describe how the bug manifests.**

Running the test script below:

```
pygame-ce 2.3.1.dev2 (SDL 2.26.5, Python 3.11.4)
first error in try/except:
Python int too large to convert to C long
<built-in function line> returned a result with an exception set

second error uncaught for trace dump:
OverflowError: Python int too large to convert to C long

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\micha\Desktop\python\bhell\coltest.py", line 28, in <module>
    pygame.draw.line(screen, 0xffffffff, SCREEN.topleft, SCREEN.bottomright)
SystemError: <built-in function line> returned a result with an exception set
```

**Expected behavior:**

No crash when high bit is set.

**Steps to reproduce:**

Test script below

**Test code**

If possible add a simple test program that shows the problem described in this report.
```
import pygame

TITLE = 'color test'
SCREEN = pygame.Rect(0, 0, 1024, 768)

pygame.init()
screen = pygame.display.set_mode(SCREEN.size, )

# These work
pygame.draw.line(screen, 'red', SCREEN.topleft, SCREEN.bottomright)
pygame.draw.line(screen, 0x00ffffff, SCREEN.topleft, SCREEN.bottomright)
pygame.draw.line(screen, 0x7fffffff, SCREEN.topleft, SCREEN.bottomright)

# These two will raise a SystemError
print('first error in try/except:')
try:
    pygame.draw.line(screen, 0x80000000, SCREEN.topleft, SCREEN.bottomright)
except SystemError as e:
    print(e.__cause__)
    print(e)

print()
print('second error uncaught for trace dump:')
pygame.draw.line(screen, 0xffffffff, SCREEN.topleft, SCREEN.bottomright)
```

**Stack trace/error output/other error logs**

See above at "manifests"

### PR DESCRIPTION
**Pull request by ankith26**
The goal of this PR is to fix #2336 and fix #1408 (the original)

The core issues are that a color is actually `Uint32`, but `long` on windows is actually `Sint32`. This leads to
- Functions returning a mapped int can return a negative value (values above and including `0x80000000` are casted incorrectly when they are `Sint32`)
- Functions expecting a mapped int can sometimes handle this negative value
- Functions expecting a mapped int can sometimes fail to handle a proper mapped int above `0x80000000`

The route I'm taking in this PR that involves least user-facing changes and breakages is
- No changes to functions returning mapped ints (maybe this could/should be changed, but I'd like to defer that to a future PR)
- All functions handling mapped int passes through the same internal C API, and this should handle *both* negative and positive-but-bigger-than-0x80000000 values, and also cap further overflows and error correctly

In the process, I figured color handling API was in a bit of a mess. We don't handle `str` and `int` colors in a lot of places. But in this PR I did not want to change anything that can have a user-facing impact (apart from the bug). This bug is related to issues in the python 2 to 3 port.

I introduced unified API for color handling (in a way that doesn't change anything userfacing)

Here is a list of old API, with new API equivalents
- `pg_RGBAFromColorObj(...)` (now removed) -> `pg_RGBAFromObjEx(..., PG_COLOR_HANDLE_SIMPLE)`
- `pg_RGBAFromObj(...)` -> `pg_RGBAFromObjEx(..., PG_COLOR_HANDLE_SIMPLE)`
- `pg_RGBAFromFuzzyColorObj(...)` (now removed) -> `pg_RGBAFromObjEx(..., PG_COLOR_HANDLE_ALL)`

A new addition `pg_MappedColorFromObj` replaces copypasted block of code that handled int incorrectly (this is what fixes this bug)

The final goal is to get rid of these temporary flags entirely, as every function should handle all `ColorValue`s. But that involves writing more docs and tests, so it's coming in a future PR.

### CODE REVIEW

**Review by zoldalma999 — COMMENTED** [review]
Overall, I like this pr, makes handling color-like objects easier. Left a few questions.

Also, when we planning to change `PG_COLOR_HANDLE_SIMPLE` calls to handle everything?

File: src_c/transform.c [hunk 1]
```diff
@@ -1766,17 +1766,10 @@ _color_from_obj(PyObject *color_obj, SDL_PixelFormat *format,
...
-
-        if (PyLong_Check(color_obj))
-            *color = (Uint32)PyLong_AsLong(color_obj);
-        else if (PyLong_Check(color_obj))
-            *color = (Uint32)PyLong_AsUnsignedLong(color_obj);
-        else if (pg_RGBAFromColorObj(color_obj, rgba_color))
-            *color = SDL_MapRGBA(format, rgba_color[0], rgba_color[1],
-                                 rgba_color[2], rgba_color[3]);
-        else
+        if (!pg_MappedColorFromObj(color_obj, format, color,
```
Comment (zoldalma999) [inline src_c/transform.c:1769]: This might be for a different PR/issue, but is there a way to make it clear in the docs and stubs that the int here means mapped int, and not a "normal" int? Do we even bother with that, do users need to worry about things like this? This goes for all uses of `pg_MappedColorFromObj`

File: src_c/pixelarray.c [hunk 2]
```diff
@@ -1546,29 +1546,11 @@ _get_subslice(PyObject *op, Py_ssize_t length, Py_ssize_t *start,
     }
     else if (PyLong_Check(op)) {
         /* Plain index: array[x, */
-        *start = PyLong_AsLong(op);
```
[the remark this answers is not in the document]
Comment (ankith26) [reply src_c/pixelarray.c:1549]: This is sorta related in the sense that it has the same double-long-check issue. This could definitely be moved to a separate PR but either way I don't think it's a big deal

File: docs/reST/c_api/base.rst [hunk 3]
```diff
@@ -91,6 +91,8 @@ C header: src_c/include/pygame.h
 
 .. c:function:: int pg_RGBAFromObj(PyObject *obj, Uint8 *RGBA)
```
[the remark this answers is not in the document]
Comment (ankith26) [reply docs/reST/c_api/base.rst:92]: I became too lazy towards the end of finishing this PR + the actual removal would lead to more diffs (as I'd have to modify some docs, there is also an example written on this function! Removing this would also lead to large diffs in slots and stuff)

File: src_c/_pygame.h [hunk 4]
```diff
@@ -321,6 +321,26 @@ struct pgColorObject {
     Uint8 len;
 };
 
+/* The positions of elements in this enum are significant, as they are used as
+ * flags with bitwise operations */
```
[the remark this answers is not in the document]
Comment (ankith26) [reply src_c/_pygame.h:325]: I don't really have a preference for either way, but may as well roll with your suggestion

File: src_c/color.c [hunk 5]
```diff
@@ -334,24 +332,22 @@ _get_double(PyObject *obj, double *val)
 }
 
 static int
-_get_color(PyObject *val, Uint32 *color)
+_get_color_component(PyObject *val, Uint8 *color)
```
[the remark this answers is not in the document]
Comment (zoldalma999) [reply src_c/color.c:335]: Yeah, that does sound better

File: docs/reST/c_api/base.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (zoldalma999) [reply docs/reST/c_api/base.rst:92]: Sounds like a follow-up PR

File: src_c/pixelarray.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (zoldalma999) [reply src_c/pixelarray.c:1549]: Sure

**Comment by MyreMylar** [comment]
For whatever reason CircleCI here is trying to build with python 3.6 - which we don't support - so it is failing.

**Review by MyreMylar — APPROVED** [review]
LGTM 👍 

The new API makes sense, and the color tests run locally for me. I also couldn't find any lingering usages of the old ways.

**Review by dr0id — APPROVED** [review]
removing code duplication is always good and makes handling colors more consistent.

LGTM

File: docs/reST/c_api/color.rst [hunk 6]
```diff
@@ -36,10 +36,16 @@ Header file: src_c/include/pygame.h
    Argument *length* must be between ``1`` and ``4`` inclusive.
    On failure, raise a Python exception and return ``NULL``.
 
-.. c:function int pg_RGBAFromColorObj(PyObject *color, Uint8 rgba[])
+.. c:function:: int pg_RGBAFromObjEx(PyObject *color, Uint8 rgba[], pgColorHandleFlags handle_flags)
```
Comment (yunline) [inline docs/reST/c_api/color.rst:39]: Maybe a description of `pgColorHandleFlags` is needed here or some where in this file (?

Comment (ankith26) [reply docs/reST/c_api/color.rst:39]: That's a good suggestion. But well, eventually the goal would be to remove `pgColorHandleFlags` entirely, so I'm not sure if it's worth documenting in this case
